using resamplefd to auto-unmask vfio INTx interrupt

zsdaka opened the pull request:
Fixes: #4999 
Current implementation is only for x86_64 and KVM.

rbradford commented:
Please follow the commit style of the project for your commit messages.

rbradford reviewed (commented):
Thanks for your PR!

I'm a bit confused - I thought with CH we exclusively ran in split mode. https://github.com/cloud-hypervisor/cloud-hypervisor/blob/b54ce6c3dbdc98ae103c38e9787e86276d99b88c/vmm/src/vm.rs#L850

When does it not run like that? Why do you need all the extra code around the split bool?

zsdaka commented:
> 



> 

If CH is not run in split mode  one day, the code here does not need to change.

rbradford commented:
CH will always run in split mode; there is no need to add complexity to handle something hypothetical in the future

zsdaka commented:
> CH will always run in split mode; there is no need to add complexity to handle something hypothetical in the future

OK，I get it.

likebreath reviewed (commented):
Thank you for the PR. Couple of quick comments:
1. Please rebase on the current "main" branch
2. Please also remove the current way of unmasking INTx via `read_bar` and `write_bar`

One more note on the commit message title: different components should be separated by `, `, say `devices, hypervisor, vmm: ...`. If these are way many components changed in one commit, you can think if it is possible to break that single commit into multiple ones, or use `misc:` as  a last resort.

I will approve the CI/workflow to run after you address the above comments. I will also take a closer look at the PR in the coming few days.

zsdaka commented:
>

> Thank you for the PR. Couple of quick comments:
> 
> 1. Please rebase on the current "main" branch
> 2. Please also remove the current way of unmasking INTx via `read_bar` and `write_bar`
> 
> One more note on the commit message title: different components should be separated by `, `, say `devices, hypervisor, vmm: ...`. If these are way many components changed in one commit, you can think if it is possible to break that single commit into multiple ones, or use `misc:` as a last resort.
> 
> I will approve the CI/workflow to run after you address the above comments. I will also take a closer look at the PR in the coming few days.

Ok. Thank you for your time and I'm sorry for my  commit style.

zsdaka commented:
I have adjusted my patches:
1. Rebased to the latest main branch.
2. Removed the INTx handling code in read/write_bar functions.
3. Removed the split mode checking code.
Now, these two patches mainly do two things:
1. Enable resample eventfd in enable_intx() and use device.set_irq_resample_fd() to register the resamplefd with the kernel VFIO.
2. The user-space ioapic EOI writes this resamplefd to trigger the INTx unmask in kernel VFIO.

likebreath commented on vmm/src/interrupt.rs:
```diff
@@ -165,6 +205,26 @@ impl InterruptSourceGroup for MsiInterruptGroup {
         None
     }
 
+    #[cfg(target_arch = "x86_64")]
+    fn enable_resamplefd(
+        &self,
+        index: InterruptIndex,
+    ) -> vm_device::interrupt::Result<()> {
```
Please address the clippy issue here with `cargo fmt`.

zsdaka replied:
Ok. I renamed the function name but forgot to run "cargo fmt".

zsdaka replied:
> Ok. I renamed the function name but forgot to run "cargo fmt".

@rbradford already done.

likebreath reviewed (commented):
I added a debug commit so that the PR will run our VFIO test. A quick question: how did you test the patch in your environment?

likebreath commented on Jenkinsfile:
```diff
@@ -342,7 +342,6 @@ pipeline {
                     when {
                         beforeAgent true
                         allOf {
-                            branch 'main'
```
@rbradford Looks like the VFIO worker is still not running with my change. May be it is a privilege thing that imposes on the PR owner instead of committer? Any advices to enable the VFIO worker temporarily here?

rbradford replied:
Jenkins ignores changes to Jenkinsfile in PRs from users who don't have write access to the repository. Jenkins doesn't really know about separate commits and anyway the committer can be easily impersonated. If you want to test on the VFIO worker you would need to open you own PR.

zsdaka commented:
> I added a debug commit so that the PR will run our VFIO test. A quick question: how did you test the patch in your environment?
1. Boot guest kernel with "pci=nomsi" kernel parameter. (Currently, CH virtio devices does not support INTx interrupt, I implement a demo version just for test.)
2. Disable HAVE_KVM_IRQ_BYPASS host kernel config, otherwise vfio device will use posted interrupt 
With the above two configs, I can see that the guest kernel can correctly handle the INTx interrupt of the device.

likebreath commented on vm-device/src/interrupt/mod.rs:
```diff
@@ -141,6 +141,27 @@ pub trait InterruptSourceGroup: Send + Sync {
+    /// interrupt automatically without userspace intervention.
+    #[allow(unused_variables)]
+    #[cfg(target_arch = "x86_64")]
+    fn unmask_notifier(&self, index: InterruptIndex) -> Option<EventFd> {
```
Question (rather than a suggestion): why not name it `resamplefd_notifier()`? Do you see the need of other "notifiers" for unmask other than using `resamplefd`?

zsdaka replied:
This interface is only for resamplefd, I think resamplefd_notifier makes more sense.

zsdaka replied:
> This interface is only for resamplefd, I think resamplefd_notifier makes more sense.

@rbradford I have renamed the function name into resamplefd_notifier.

likebreath commented on vm-device/src/interrupt/mod.rs:
```diff
@@ -141,6 +141,27 @@ pub trait InterruptSourceGroup: Send + Sync {
     #[allow(unused_variables)]
     fn notifier(&self, index: InterruptIndex) -> Option<EventFd>;
 
+    /// Enable the resamplefd.
+    /// With the resamplefd, vfio INTx interrupt can be automatically unamsk in EOI
+    #[cfg(target_arch = "x86_64")]
+    fn enable_resamplefd(&self, _index: InterruptIndex) -> Result<()> {
+        // Not all interrupt sources need resamplefd.
+        // If not, we can have a no-op here.
+        Ok(())
+    }
+
+    /// Returns an interrupt resample notifier from this interrupt.
+    ///
+    /// An interrupt resample notifier allows for kvm re-enable level triggered
+    /// interrupt automatically without userspace intervention.
+    #[allow(unused_variables)]
+    #[cfg(target_arch = "x86_64")]
+    fn unmask_notifier(&self, index: InterruptIndex) -> Option<EventFd> {
+        // Not all interrupt sources have a unmask notifier.
+        // If not, we can return None here.
+        None
+    }
+
```
These new APIs are making sense to me in terms of supporting `resamplefd` for INTx with split irqchip. 
@sboeuf Do you think this is the right extension `InterruptSourceGroup` (from the `vm-device` crate perspective)?

likebreath commented on vmm/src/interrupt.rs:
```diff
@@ -23,6 +23,10 @@ struct InterruptRoute {
     irq_fd: EventFd,
     registered: AtomicBool,
+    #[cfg(target_arch = "x86_64")]
+    resample_fd: Mutex<Option<EventFd>>,
```
I think the related code can be simplified if you use `resample_fd: EventFd`. 

You just need to initialize it from the constructor function e.g. `InterruptRoute::new()`, and use `resample_enabled` to decide whether a `unmask_notifier()` will return `None` or `Some`.

zsdaka replied:
Currently, only VFIO devices will use resamplefd. If we create the resamplefd in the new() function, then all the IOAPIC and MSI/MSIX interrupts will have a resamplefd which is unnecessary. Therefore, I added a new interface called enable_resamplefd, which only creates the resamplefd when it is needed.

likebreath replied:
Actually, my confusion was from using both `Mutex` or the `AtomicBool` at the same time.

IIUC, we only need to avoid racing for initializing the `resample_fd` (which is the only "write") and all following accesses are "read"(e.g. `clone()`). So I think either one of the following would work, right?
```
    #[cfg(target_arch = "x86_64")]
    resample_fd: Mutex<Option<EventFd>>,
```

or 
```
    #[cfg(target_arch = "x86_64")]
    resample_fd: Option<EventFd>,
    #[cfg(target_arch = "x86_64")]
    resample_enabled: AtomicBool,
```

zsdaka replied:
Thank you very much for your correction, I think you are right.

zsdaka replied:
> Actually, my confusion was from using both `Mutex` or the `AtomicBool` at the same time.
> 
> IIUC, we only need to avoid racing for initializing the `resample_fd` (which is the only "write") and all following accesses are "read"(e.g. `clone()`). So I think either one of the following would work, right?
> 
> ```
>     #[cfg(target_arch = "x86_64")]
>     resample_fd: Mutex<Option<EventFd>>,
> ```
> 
@rbradford I have used a Mutex to serialize access and clean the AtomicBool variable in the latest patch.

> or
> 
> ```
>     #[cfg(target_arch = "x86_64")]
>     resample_fd: Option<EventFd>,
>     #[cfg(target_arch = "x86_64")]
>     resample_enabled: AtomicBool,
> ```

likebreath commented on pci/src/vfio_user.rs:
```diff
@@ -391,6 +391,47 @@ impl Vfio for VfioUserClientWrapper {
+            irq_index,
+            event_rfds.len()
+        );
+        let fds: Vec<i32> = event_rfds.iter().map(|e| e.as_raw_fd()).collect();
+
+        // Batch into blocks of 16 fds as sendmsg() has a size limit
+        let mut sent_fds = 0;
+        let num_fds = event_rfds.len() as u32;
+        while sent_fds < num_fds {
+            let remaining_fds = num_fds - sent_fds;
+            let count = if remaining_fds > 16 {
+                16
+            } else {
+                remaining_fds
+            };
+
+            self.client
+                .lock()
+                .unwrap()
+                .set_irqs(
+                    irq_index,
+                    VFIO_IRQ_SET_DATA_EVENTFD | VFIO_IRQ_SET_ACTION_UNMASK,
+                    sent_fds,
+                    count,
+                    &fds[sent_fds as usize..(sent_fds + count) as usize],
+                )
+                .map_err(VfioError::VfioUser)?;
+
+            sent_fds += count;
+        }
+
```
This part can be simplified a lot. There will always only be one resamplefd, so you don't need the loop here.

zsdaka replied:
> This part can be simplified a lot. There will always only be one resamplefd, so you don't need the loop here.

@rbradford I have removed the loop in the latest patch.

likebreath reviewed (commented):
Overall looks good. Some comments below. 

Please drop my last commit as it does not work. I will create a test PR to run the VFIO CI after these comments are addressed. Thank you.

likebreath commented:
> > I added a debug commit so that the PR will run our VFIO test. A quick question: how did you test the patch in your environment?
> 
> 1. Boot guest kernel with "pci=nomsi" kernel parameter. (Currently, CH virtio devices does not support INTx interrupt, I implement a demo version just for test.)
> 2. Disable HAVE_KVM_IRQ_BYPASS host kernel config, otherwise vfio device will use posted interrupt
>    With the above two configs, I can see that the guest kernel can correctly handle the INTx interrupt of the device.

Cool. Sounds like that's not something we can do in a new integration test (or extension to our current `test_vfio` test)? @sboeuf @rbradford

zsdaka commented on vmm/src/interrupt.rs:
```diff
@@ -82,6 +92,36 @@ impl InterruptRoute {
+
+    #[cfg(target_arch = "x86_64")]
+    pub fn unmask_notifier(&self) -> Option<EventFd> {
+        if let Some(rfd) = &*self.resample_fd.lock().unwrap() {
```
OK，got it.

likebreath commented on vm-device/src/interrupt/mod.rs:
```diff
@@ -141,6 +141,27 @@ pub trait InterruptSourceGroup: Send + Sync {
 
+    /// Enable the resamplefd.
+    /// With the resamplefd, vfio INTx interrupt can be automatically unamsk in EOI
+    #[cfg(target_arch = "x86_64")]
+    fn enable_resamplefd(&self, _index: InterruptIndex) -> Result<()> {
+        // Not all interrupt sources need resamplefd.
+        // If not, we can have a no-op here.
+        Ok(())
+    }
+
```
Do we need the counter part for `disable_resamplefd()`? If not, why?

zsdaka replied:
[https://github.com/torvalds/linux/blob/master/drivers/vfio/pci/vfio_pci_intrs.c]()
```
static void vfio_intx_disable(struct vfio_pci_core_device *vdev)
{
	vfio_virqfd_disable(&vdev->ctx[0].unmask);
	vfio_virqfd_disable(&vdev->ctx[0].mask);
	vfio_intx_set_signal(vdev, -1);
	vdev->irq_type = VFIO_PCI_NUM_IRQS;
	vdev->num_ctx = 0;
	kfree(vdev->ctx);
}
```
The VFIO kernel will stop polling the resamplefd when we call device.disable_irq() from CH. However, the CH user-space IOAPIC may still trigger the resamplefd, although this is harmless. Therefore, adding a disable_resamplefd() function would be more reasonable. I will add this interface and call it from the device.disable_intx() function.

zsdaka replied:
> Do we need the counter part for `disable_resamplefd()`? If not, why?

@rbradford I have added the disable_resamplefd() function, which will be called from the device.disable_intx() function.

likebreath commented on pci/src/vfio.rs:
```diff
@@ -886,11 +908,26 @@ impl VfioCommon {
                     return Err(VfioPciError::MissingNotifier);
                 }
+
+                #[cfg(target_arch = "x86_64")]
+                {
+                    intx.interrupt_source_group
+                        .enable_resamplefd(0)
+                        .map_err(VfioPciError::EnableResamplefd)?;
+
+                    if let Some(reventfd) = intx.interrupt_source_group.unmask_notifier(0) {
+                        self.vfio_wrapper
+                            .set_irq_resample_fd(VFIO_PCI_INTX_IRQ_INDEX, vec![&reventfd])
+                            .map_err(VfioPciError::EnableIntx)?;
+                    } else {
+                        return Err(VfioPciError::MissingResampleNotifier);
+                    }
+                }
```
Do we need to do the counter part for `disable_intx()`?

zsdaka replied:
The same as the above question. I will add the counter part.

zsdaka replied:
> The same as the above question. I will add the counter part.

@rbradford already done

likebreath commented on devices/src/ioapic.rs:
```diff
@@ -399,8 +399,16 @@ impl InterruptController for Ioapic {
             let entry = &mut self.reg_entries[i];
-            // Clear Remote IRR bit
             if vector(*entry) == vec && trigger_mode(*entry) == 1 {
+                // resamplefd write will be skipped due to split irqchip mode,
+                // so userspace EOI should handle the resamplefd write instead.
+                if let Some(rfd) = self
+                    .interrupt_source_group
+                    .unmask_notifier(i as InterruptIndex)
+                {
+                    rfd.write(1).unwrap();
+                }
```
I wonder if there is a chance the same interrupt fired while processing this `eoi`?

likebreath replied:
Found related comments and code from crosvm: https://github.com/google/crosvm/blob/main/devices/src/irqchip/ioapic.rs#L281-L286

zsdaka replied:
> I wonder if there is a chance the same interrupt fired while processing this `eoi`?

The hardware device's interrupt handler, vfio_intx_handler(), will mask the device's INTx interrupt, no interrupt from this device will be fired until the resamplefd is triggered. Once the resamplefd is triggered, the vfio_pci_intx_unmask_handler will be called to unmask the device's INTx interrupt again. So, before the "resamplefd" function is triggered, we will not see any interrupts. Please let me know if there are any errors in my understanding.

likebreath reviewed (commented):
Added a couple more questions.

likebreath commented:
@rbradford  and I discussed about this issue earlier today in our [community meeting](https://github.com/cloud-hypervisor/cloud-hypervisor#slack). Our major concern is that we won't have a proper way to test this feature in our CI environment. Carrying some untested feature is something we try to avoid, particularly for complex features like this that touch many components of the code base. Meanwhile, we believe there will be rarely a real use case for this feature considering all modern devices are MSI/MSI-X compatible. 

@zsdaka So we decided to not include this feature upstream at this point. Sorry about the long turnaround time. I appreciate your time and contributions. Any questions or comments, please let me know.

zsdaka commented:
> @rbradford and I discussed about this issue earlier today in our [community meeting](https://github.com/cloud-hypervisor/cloud-hypervisor#slack). Our major concern is that we won't have a proper way to test this feature in our CI environment. Carrying some untested feature is something we try to avoid, particularly for complex features like this that touch many components of the code base. Meanwhile, we believe there will be rarely a real use case for this feature considering all modern devices are MSI/MSI-X compatible.
> 
> @zsdaka So we decided to not include this feature upstream at this point. Sorry about the long turnaround time. I appreciate your time and contributions. Any questions or comments, please let me know.

@likebreath Your concern is reasonable. I am glad to participate in the CH community and also appreciate everyone taking the time to review my code.
